torch/_lobpcg.py

```diff
@@ -0,0 +1,721 @@
+    if B is not None:
+        # A and B must have the same shapes:
+        assert A.shape == B.shape, (A.shape, B.shape)
+
+    dtype = _utils.get_floating_dtype(A)
+    device = A.device
+    feps = {torch.float32: 1.2e-07,
+            torch.float64: 2.23e-16}[dtype]
+    tol = feps ** 0.5
+
+    m = A.shape[-1]
+    k = (1 if X is None else X.shape[-1]) if k is None else k
+    n = (k if n is None else n) if X is None else X.shape[-1]
+
+    if (m < 3 * n):
+        raise ValueError(
+            'LPBPCG algorithm is not applicable when the number of A rows (={})'
+            ' is smaller than 3 x the number of requested eigenpairs (={})'
+            .format(m, n))
+
+    method = 'ortho' if method is None else method
+
+    iparams = {
+        'm': m,
+        'n': n,
+        'k': k,
+        'niter': 1000 if niter is None else niter,
+    }
+
+    fparams = {
+        'tol': tol,
+    }
+
+    bparams = {
+        'largest': True if largest is None else largest
+    }
+
+    if method == 'ortho':
+        if ortho_iparams is not None:
+            iparams.update(ortho_iparams)
+        if ortho_fparams is not None:
+            fparams.update(ortho_fparams)
+        if ortho_bparams is not None:
+            bparams.update(ortho_bparams)
+        iparams['ortho_i_max'] = iparams.get('ortho_i_max', 3)
+        iparams['ortho_j_max'] = iparams.get('ortho_j_max', 3)
+        fparams['ortho_tol'] = fparams.get('ortho_tol', tol)
+        fparams['ortho_tol_drop'] = fparams.get('ortho_tol_drop', tol)
+        fparams['ortho_tol_replace'] = fparams.get('ortho_tol_replace', tol)
+        bparams['ortho_use_drop'] = bparams.get('ortho_use_drop', False)
+
+    if len(A.shape) > 2:
+        N = int(torch.prod(torch.tensor(A.shape[:-2])))
+        bA = A.reshape((N,) + A.shape[-2:])
+        bB = B.reshape((N,) + A.shape[-2:]) if B is not None else None
+        bX = X.reshape((N,) + X.shape[-2:]) if X is not None else None
+        bE = torch.empty((N, k), dtype=dtype, device=device)
+        bXret = torch.empty((N, m, k), dtype=dtype, device=device)
+        for i in range(N):
+            A_ = bA[i]
+            B_ = bB[i] if bB is not None else None
+            X_ = torch.randn((m, n), dtype=dtype, device=device) if bX is None else bX[i]
+            assert len(X_.shape) == 2 and X_.shape == (m, n), (X_.shape, (m, n))
+            iparams['batch_index'] = i
+            worker = LOBPCG(A_, B_, X_, iK, iparams, fparams, bparams, method, tracker)
+            worker.run()
+            bE[i] = worker.E[:k]
+            bXret[i] = worker.X[:, :k]
+        return bE.reshape(A.shape[:-2] + (k,)), bXret.reshape(A.shape[:-2] + (m, k))
+
+    X = torch.randn((m, n), dtype=dtype, device=device) if X is None else X
+    assert len(X.shape) == 2 and X.shape == (m, n), (X.shape, (m, n))
+
+    worker = LOBPCG(A, B, X, iK, iparams, fparams, bparams, method, tracker)
+    worker.run()
+    return worker.E[:k], worker.X[:, :k]
+
+
+@torch.jit.script
```

Comment:
I'm noticing that applying this decorator to a whole class isn't really used in the rest of the code base. Usage of `torch.jit.script` outside of `torch/jit/` and testing code is rare. So just flagging it for a maintainer to check.

Reply:
Decorating LOBPCG class with `torch.jit.script` is required for `torch.jit.script(torch.lobpcg)` to work: the current implementation overrides the `call_tracker` method in a non-standard way, see
https://github.com/Quansight/pytorch/blob/pearu/lowrank-linalg/torch/_lobpcg.py#L719-L721
(because TorchScript does not support function arguments). This override is effective only in Python mode and enables users to provide a callback function for tracking the iterations.

Reply:
@ezyang -- can you comment on whether the function should be decorated or not?

Reply:
I would recommend not decorating by default.

@ezyang -- thoughts?